hkmc2/shared/src/main/scala/hkmc2/bbml/ConstraintSolver.scala

```diff
@@ -87,16 +87,17 @@ class ConstraintSolver(infVarState: InfVarUid.State, elState: Elaborator.State,
         else
           val bd = if v.lvl >= rest.lvl then rest else extrude(rest)(using v.lvl, true, mutable.HashMap.empty)
           if pol then
-            val nc = Type.mkNegType(bd)
+            val nc = Type.mkNegType(bd).toDnf
             log(s"New bound: ${v.showDbg} <: ${nc.showDbg}")
             cctx.nest(v -> nc) givenIn:
               v.state.upperBounds ::= nc
               v.state.lowerBounds.foreach(lb => constrainImpl(lb, nc))
           else
-            log(s"New bound: ${v.showDbg} :> ${bd.showDbg}")
-            cctx.nest(bd -> v) givenIn:
-              v.state.lowerBounds ::= bd
-              v.state.upperBounds.foreach(ub => constrainImpl(bd, ub))
+            val c = bd.toDnf
+            log(s"New bound: ${v.showDbg} :> ${c.showDbg}")
+            cctx.nest(c -> v) givenIn:
```

Comment:
I just noticed you're using a purely functional cache. Doing so has the wrong complexity (exponential instead of quadratic). Are we doing this to mirror the formally described system in the paper?

Reply:
Yes, we use a purely functional cache in the paper. But we can also adopt a mutable one in the implementation (actually, there is already a TODO comment above the `CCtx` class for this).